services/mac_manager/src/main/java/com/futurewei/alcor/macmanager/pool/impl/BitmapPoolImpl.java

```diff
@@ -0,0 +1,261 @@
+/*
+ *
+ * Copyright 2019 The Alcor Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ *         you may not use this file except in compliance with the License.
+ *         You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *         Unless required by applicable law or agreed to in writing, software
+ *         distributed under the License is distributed on an "AS IS" BASIS,
+ *         WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *         See the License for the specific language governing permissions and
+ *         limitations under the License.
+ * /
+ */
+
+package com.futurewei.alcor.macmanager.pool.impl;
+
+import com.futurewei.alcor.common.db.CacheException;
+import com.futurewei.alcor.common.db.CacheFactory;
+import com.futurewei.alcor.common.db.DistributedLockFactory;
+import com.futurewei.alcor.common.db.IDistributedLock;
+import com.futurewei.alcor.common.exception.DistributedLockException;
+import com.futurewei.alcor.macmanager.dao.MacRangeMappingRepository;
+import com.futurewei.alcor.macmanager.dao.MacRangePartitionRepository;
+import com.futurewei.alcor.macmanager.dao.MacStateRepository;
+import com.futurewei.alcor.macmanager.exception.MacAddressFullException;
+import com.futurewei.alcor.macmanager.exception.MacAddressRetryLimitExceedException;
+import com.futurewei.alcor.macmanager.exception.MacRepositoryTransactionErrorException;
+import com.futurewei.alcor.macmanager.pool.MacPoolApi;
+import com.futurewei.alcor.macmanager.utils.MacManagerConstant;
+import com.futurewei.alcor.web.entity.mac.MacRange;
+import com.futurewei.alcor.web.entity.mac.MacRangePartition;
+import org.apache.commons.lang3.RandomUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.util.BitSet;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.futurewei.alcor.macmanager.utils.MacUtils.*;
+
+//@Component
+public class BitmapPoolImpl implements MacPoolApi {
+
+    private static final Logger logger = LoggerFactory.getLogger(BitmapPoolImpl.class);
+
+    private static final int PARTITION_SIZE = 4096;
+
+    @Value("${macmanager.retrylimit}")
+    private long nRetryLimit;
+
+    @Autowired
+    private MacRangePartitionRepository macRangePartitionRepository;
+
+    @Autowired
+    private MacRangeMappingRepository macRangeMappingRepository;
+
+    private final IDistributedLock lock;
+
+    @Autowired
+    public BitmapPoolImpl(DistributedLockFactory cacheFactory){
+        lock = cacheFactory.getDistributedLock(MacRangePartition.class);
+    }
+
+    @Override
+    public String allocate(String oui, MacRange macRange) throws MacAddressFullException,
```

Comment:
Going through the whole algorithm, we didn't find the usage of MacRangePartition.setBitSet. 

In the current implementation, it seems that the allocation is still using the range-level BitSet, instead of using the partition-level BitSet. If we have partition-level bitset, the range-level one should be abandoned.

Let us discuss more in next open-source meeting.